### PR TITLE
Template

### PR DESCRIPTION
**Pull request by psunnn**

### CODE REVIEW

File: .migrations/V20200123054713__initial_table_create.sql [hunk 1]
```diff
@@ -71,8 +71,7 @@ CREATE TABLE IF NOT EXISTS task (
   namespace text,
   max_cpu_used integer,
   max_memory_used integer,
-  pod_events jsonb,
-  cloudtrail_notifications jsonb
+  pod_events jsonb
```
Comment (psunnn) [inline .migrations/V20200123054713__initial_table_create.sql]: cc @ujjwalsfix 😁

File: ui/src/components/App.tsx [hunk 2]
```diff
@@ -58,6 +58,18 @@ class App extends React.Component<ConnectedProps<typeof connector>> {
             <Route exact path="/tasks/create" component={CreateTaskForm} />
             <Route path="/tasks/:definitionID" component={Task} />
             <Route path="/tasks/alias/:alias" component={Task} />
+
+            <Route
+              exact
+              path="/templates"
+              component={() => <span>tpl list</span>}
```
Comment (psunnn) [inline ui/src/components/App.tsx]: wip

File: flotilla/endpoints.go [hunk 3]
```diff
@@ -95,6 +97,8 @@ func (ep *endpoints) decodeListRequest(r *http.Request) listRequest {
 
 	lr.limit, _ = strconv.Atoi(ep.getURLParam(params, "limit", "1024"))
 	lr.offset, _ = strconv.Atoi(ep.getURLParam(params, "offset", "0"))
+
+	// @TODO not every entity that can be listed has a group name...
 	lr.sortBy = ep.getURLParam(params, "sort_by", "group_name")
```
Comment (psunnn) [inline flotilla/endpoints.go]: TODO: should probably add another method called decodeListRequestV7 that doesn't assume every entity has a group name field

File: flotilla/endpoints.go [hunk 4]
```diff
@@ -381,14 +385,7 @@ func (ep *endpoints) CreateRunV2(w http.ResponseWriter, r *http.Request) {
 		ep.encodeResponse(w, run)
 	}
 }
-func stringInSlice(a string, list []string) bool {
```
Comment (psunnn) [inline flotilla/endpoints.go]: see the utils package

File: services/execution.go [hunk 5]
```diff
@@ -261,7 +204,12 @@ func (es *executionService) constructBaseRunFromExecutable(executable state.Exec
 	// Compute the executable command based on the execution request. If the
 	// execution request did not specify an overriding command, use the computed
 	// `executableCmd` as the Run's Command.
-	executableCmd := executable.GetExecutableCommand(req)
+	executableCmd, err := executable.GetExecutableCommand(req)
```
Comment (psunnn) [inline services/execution.go]: note: this function, `constructBaseRunFromExecutable` is invoked by both `es.CreateTemplateRun...` and `es.CreateDefinitionRun...`. Both structs implement the Executable interface, but `GetExecutableCommand` for a template is more complicated

File: services/execution.go [hunk 6]
```diff
@@ -474,3 +422,120 @@ func (es *executionService) Terminate(runID string, userInfo state.UserInfo) err
 func (es *executionService) ListClusters() ([]string, error) {
 	return es.ecsClusterClient.ListClusters()
 }
+
+//
+// sanitizeExecutionRequestCommonFields does what its name implies - sanitizes
+// several common request fields (mostly around ECS/EKS differences).
+//
+func (es *executionService) sanitizeExecutionRequestCommonFields(fields *state.ExecutionRequestCommon, privileged bool) {
```
Comment (psunnn) [inline services/execution.go]: keeping it dry

File: state/models.go [hunk 7]
```diff
@@ -734,3 +737,179 @@ func (w *Record) Equal(other Record) bool {
...
+	et := ExecutableTypeTemplate
+	return &et
+}
+
+func (t Template) GetExecutableResources() *ExecutableResources {
+	return &t.ExecutableResources
+}
+
+func (t Template) GetExecutableCommand(req ExecutionRequest) (string, error) {
+	var (
```
Comment (psunnn) [inline state/models.go]: note: this is where jsonschema validation and go template strings come into play - deserves a closer look

File: state/models.go [hunk 8]
```diff
@@ -734,3 +737,179 @@ func (w *Record) Equal(other Record) bool {
...
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("tpl-%s", uuid4), nil
+}
+
+func (t *Template) IsValid() (bool, []string) {
+	// TODO: validate that json schema can be dumped into the command template
+	// with no <no value> values.
+	return false, []string{}
```
Comment (psunnn) [inline state/models.go]: todo: needs to be implemented

File: state/pg_state_manager.go [hunk 9]
```diff
@@ -954,3 +968,174 @@ func (e CloudTrailNotifications) Value() (driver.Value, error) {
 	res, _ := json.Marshal(e)
 	return res, nil
 }
+
```
Comment (psunnn) [inline state/pg_state_manager.go]: can we please switch to gorm

File: worker/retry_worker.go [hunk 10]
```diff
@@ -67,7 +67,9 @@ func (rw *retryWorker) runOnce() {
 
 	runList, err := rw.sm.ListRuns(25, 0, "started_at", "asc", map[string][]string{"status": {state.StatusNeedsRetry}}, nil, engines)
 
-	rw.log.Log("message", fmt.Sprintf("Got %v jobs to retry", runList.Total))
+	if runList.Total > 0 {
```
Comment (psunnn) [inline worker/retry_worker.go]: trying to reduce the log output here

File: worker/status_worker.go [hunk 11]
```diff
@@ -90,30 +90,33 @@ func (sw *statusWorker) runOnceEKS() {
...
+				sw.processEKSRun(run)
+				sw.processEKSRunMetrics(run)
 			}
 		}
 	}
 }
 func (sw *statusWorker) acquireLock(run state.Run, purpose string, expiration time.Duration) bool {
 	set, err := sw.redisClient.SetNX(fmt.Sprintf("%s-%s", run.RunID, purpose), sw.workerId, expiration).Result()
 	if err != nil {
-		_ = sw.log.Log("message", "unable to set lock", "error", fmt.Sprintf("%+v", err))
```
Comment (psunnn) [inline worker/status_worker.go]: trying to reduce the log output here

File: state/models.go [hunk 12]
```diff
@@ -734,3 +737,179 @@ func (w *Record) Equal(other Record) bool {
...
+
+func (t TemplateExecutionRequest) GetExecutionRequestCustom() *map[string]interface{} {
+	return &map[string]interface{}{
+		TemplatePayloadKey: t.TemplatePayload,
+	}
+}
+
+type Template struct {
+	TemplateID      string `json:"template_id"`
+	Type            string `json:"type"`
```
Comment (ujjwalsfix) [inline state/models.go]: maybe we should use `template_name` - similar to `template_id` `type` tends to be a keyword in few different languages.

Comment (psunnn) [reply state/models.go]: 👍

File: services/execution.go [hunk 13]
```diff
@@ -474,3 +422,120 @@ func (es *executionService) Terminate(runID string, userInfo state.UserInfo) err
...
+
+func (es *executionService) constructRunFromTemplate(template state.Template, req *state.TemplateExecutionRequest) (state.Run, error) {
+	run, err := es.constructBaseRunFromExecutable(template, req)
+
+	if err != nil {
+		return run, err
+	}
+
+	// @TODO: how to handle these.
+	run.DefinitionID = template.TemplateID
```
Comment (ujjwalsfix) [inline services/execution.go]: Does run.Definition_Id have a fk relationship to Definition table?

File: services/template.go [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+
+		for i, e := range prevTags {
+			if e != currTags[i] {
+				return true
+			}
+		}
+	}
+
+	if reflect.DeepEqual(prev.Schema, curr.Schema) == false {
+		fmt.Println("M")
```
Comment (ujjwalsfix) [inline services/template.go]: remove

File: services/template.go [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+	return ts.sm.ListTemplatesLatestOnly(limit, offset, sortBy, order)
+}
+
+// diff performs a diff between all fields (except for TemplateName and
+// Version) of two templates.
+func (ts *templateService) diff(prev state.Template, curr state.Template) bool {
+	if prev.TemplateName != curr.TemplateName {
+		return true
+	}
+	if prev.CommandTemplate != curr.CommandTemplate {
```
Comment (ujjwalsfix) [inline services/template.go]: Any whitespace related issues with writing to db and back?

File: services/template.go [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+	return false
+}
+
+// constructTemplateFromCreateTemplateRequest takes a CreateTemplateRequest and
+// dumps the requisite fields into a Template.
+func (ts *templateService) constructTemplateFromCreateTemplateRequest(req *state.CreateTemplateRequest) (state.Template, error) {
+	tpl := state.Template{}
+
+	if len(req.TemplateName) > 0 {
+		tpl.TemplateName = req.TemplateName
```
Comment (ujjwalsfix) [inline services/template.go]: Do we want to enforce any rules here on naming?

File: services/template.go [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+	if req.Cpu != nil {
+		tpl.Cpu = req.Cpu
+	}
+	if req.Env != nil {
+		tpl.Env = req.Env
+	}
+	if req.Privileged != nil {
+		tpl.Privileged = req.Privileged
+	}
+	if req.AdaptiveResourceAllocation != nil {
```
Comment (ujjwalsfix) [inline services/template.go]: What is the default behavior when not set?

File: state/models.go [hunk 18]
```diff
@@ -734,3 +737,137 @@ func (w *Record) Equal(other Record) bool {
...
+	return t.TemplateID
+}
+
+// NewTemplateID returns a new uuid for a Template
+func NewTemplateID(t Template) (string, error) {
+	uuid4, err := newUUIDv4()
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("tpl-%s", uuid4), nil
```
Comment (ujjwalsfix) [inline state/models.go]: Might want to truncate first 4 characters off uuid4

Comment (ujjwalsfix) [reply state/models.go]: Keep the length standard

File: worker/submit_worker.go [hunk 19]
```diff
@@ -78,34 +82,68 @@ func (sw *submitWorker) runOnce() {
...
-		}
 
-		//
-		// Only valid to process if it's in the StatusQueued state
-		//
-		if run.Status == state.StatusQueued {
+			// 2. Check for existence of run.ExecutableID; set to run.DefinitionID if
+			// not set.
+			if run.ExecutableID == nil {
+				defID := run.DefinitionID
```
Comment (ujjwalsfix) [inline worker/submit_worker.go]: is run.DefinitionID always going to be set?

File: worker/submit_worker.go [hunk 20]
```diff
@@ -78,34 +82,68 @@ func (sw *submitWorker) runOnce() {
...
+					}
+					continue
+				}
+
+				// Execute the run using the execution engine.
+				launched, retryable, err = sw.ee.Execute(d, run, sw.sm)
+				break
+			case state.ExecutableTypeTemplate:
+				var tpl state.Template
+				tpl, err = sw.sm.GetTemplateByID(*run.ExecutableID)
```
Comment (ujjwalsfix) [inline worker/submit_worker.go]: Any nil checks required here?

File: state/models.go [hunk 21]
```diff
@@ -734,3 +737,157 @@ func (w *Record) Equal(other Record) bool {
...
+}
+
+func (t *Template) IsValid() (bool, []string) {
+	conditions := []validationCondition{
+		{len(t.TemplateName) == 0, "string [template_name] must be specified"},
```
Comment (ujjwalsfix) [inline state/models.go]: Do we want to enforce any particular convention. For instance `PythonTask` or `python_task`

File: state/pg_queries.go [hunk 22]
```diff
@@ -179,3 +179,53 @@ const GetWorkerSQL = WorkerSelect + "\nwhere worker_type = $1 and engine = $2"
...
+  cpu,
+  gpu
+FROM template
+`
+
+// ListTemplatesSQL postgres specific query for listing templates
+const ListTemplatesSQL = TemplateSelect + "\n%s limit $1 offset $2"
+
+// GetTemplateByIDSQL postgres specific query for getting a single template
+const GetTemplateByIDSQL = TemplateSelect + "\nwhere template_id = $1"
```
Comment (ujjwalsfix) [inline state/pg_queries.go]: Do we want to put an index on template_id - otherwise it is a full table scan

File: state/pg_queries.go [hunk 23]
```diff
@@ -179,3 +179,53 @@ const GetWorkerSQL = WorkerSelect + "\nwhere worker_type = $1 and engine = $2"
...
+    privileged,
+    cpu,
+    gpu
+  FROM template
+  ORDER BY template_name, version DESC, template_id
+  LIMIT $1 OFFSET $2;
+`
+
+// GetTemplateLatestOnlySQL get the latest version of a specific template name.
+const GetTemplateLatestOnlySQL = TemplateSelect + "\nWHERE template_name = $1 ORDER BY version DESC LIMIT 1;"
```
Comment (ujjwalsfix) [inline state/pg_queries.go]: index on tamplte_name too

File: services/template.go (hunk 17 above)
Comment (psunnn) [reply services/template.go]: privileged set to false; ARA set to true

**Pull request merged**